sycl/doc/ControllingSYCLLanguageFeatures.md

```diff
@@ -0,0 +1,195 @@
+
+
+[SYCL_INTEL_extension_api]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ExtensionMechanism/SYCL_INTEL_extension_api.asciidoc
+[Predefined macros]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/ExtensionMechanism/SYCL_INTEL_extension_api.asciidoc#predefined-macros
+
+#### -sycl-ext vs. -fsycl-extension-name/-fno-sycl-extension-name
+
+Another one way to enable/disable extensions is to provide separate option
+for each extensions.
+
+There are pros and cons of both approaches:
+
+##### Single flag, i.e. -sycl-ext
+
+Pros:
+- We don't have to hardcode list of known extensions and update the compiler
+  for each new extension we have prepared: `-cl-std` works in this manner
+
+  - This simplifies prototyping of new extensions, especially if they are
+    header-only
+
+- This allows to easily treat header-only extensions in the same way as
+  actual compiler extension and change the implementation without any
+  updates in compiler/toolchain interface (command line options)
+
+- Enabling/disabling extension via this option could automatically define
+  corresponding macro, which can be used by header-only extensions to hide
+  new classes/APIs and other stuff introduced by the extension to avoid it
+  being accidentally used
+
+Cons:
+- Without list of known extensions, we cannot emit proper diagnostic that some
+  unknown extension was enabled just because of the typo in its spelling
+
+  - Potentially, we could introduce one more option, which will allow to check
+    that list of enabled/disabled extensions doesn't contain anything unknown,
+    but this again means change of the compiler for each even header-only
+    extension
+
+- According to [Predefined macros] section from [SYCL_INTEL_extension_api]
+  proposal, extension-specific macro should not only be defined, but also has a
+  value in particular form. How can we automatically put any meaningful value
+  in there without having predefined list of known extensions? Do we need to
+  extend the format so user can specify which version of the extension is
+  needed (`-sycl-ext=+EXTENSION_NAME1=123`)?
+
+##### Separate flag for each extension
+
+Generic form of command line option to control SYCL/DPC++ language extension
+is `-f[no-]sycl-extension-name`. For example, `-fsycl-usm` flag enables support
+for [USM] extension, `-fno-sycl-usm` disables it.
+
+[USM]: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/USM/USM.adoc
+
+Flags controlling the same extension might be set multiple times in the same
+compiler invocation: the last one overrides previous flag.
+
+Pros:
+- We will automatically get a diagnostic if user made a typo in an option name,
+  which corresponds to a particular extension
+
+Cons:
+- Seems like a significant amount of new flags coming to the compiler
+- Harder to prototype and implement new extensions
+- What about header-only extensions? Bunch of compiler options which are only
```

Comment:
The current list of extensions may be large, but once the next SYCL version is out, most of them will simply be the new "normal" under sycl-2020. I don't think is a huge burden for users to set them individually. Until there is a complete 2020 implementation, using the -sycl-std=dpcpp-07 should set all relevant not-target specific extensions, so the user should only have to add those extensions that are target specific. That list should not be that large, unless I am mistaken, so having separate flags should be fine.

Reply:
Depending on how much of dpc++ gets upstreamed to llvm, this may be a long time. I would assume, however, a large part of the changes and extensions will stabilize. There is a large gap between SYCL 1.2.1 and DPC++ that will be much smaller after SYCL 2020.